libobs/obs-source.c

```diff
@@ -2879,19 +2862,29 @@ cache_video(struct obs_source *source, const struct obs_source_frame *frame)
 
 	pthread_mutex_lock(&source->async_mutex);
 
-	if (source->async_frames.num >= MAX_ASYNC_FRAMES) {
-		free_async_cache(source);
-		source->last_frame_ts = 0;
-		pthread_mutex_unlock(&source->async_mutex);
-		return NULL;
-	}
-
 	if (async_texture_changed(source, frame)) {
 		free_async_cache(source);
 		source->async_cache_width = frame->width;
 		source->async_cache_height = frame->height;
 	}
 
+	if (source->async_frames.num >= MAX_ASYNC_FRAMES) {
```

Comment:
Since we no longer call `free_async_cache`, nothing resets `async_frames.num` once this condition is hit so it continuously executes, marking all frames as available.

Reply:
If all the frames are used, then this function returns NULL and doesn't mark anything. Are you talking about code in some other function? `available` is a local variable.

Reply:
Hmm maybe I'm misunderstanding what is supposed to be happening here. The previous code would reset the cache and async frames when we reached `MAX_ASYNC_FRAMES`. The new code would keep returning `NULL` and leaving the `async_cache` untouched until a frame was removed elsewhere, presumably blocking output until that happens. I was under the impression it would "wipe" all the frames like the old code by marking them as available again but that doesn't seem to be the case. During testing it does seem like things work OK when this is triggered, with a breakpoint for ~10 secs CPU runs at 100% for a bit then all the media sources catch up. As long as that's the intention then it seems fine.

Reply:
Retaining something like that had crossed my mind, but I figured I'd let the frames retire naturally (the "zero code" option), and it seemed to work. You said you had run into some other weirdness, and maybe that could be fixed here, but we can debug that later.

Reply:
When you say debug later, do you mean after it's merged or before? Just asking because I wanted to see whether we should merge it or not before 27.1

Reply:
I was thinking after, but I can take a look at the issue first.

Reply:
I tried the debugger pause and recover scenario, and it seems to be behaving as expected. I think this is fine to merge.

Reply:
This broke streams in Media Source. I've opened the issue before thinking of searching for the matching pull request.

https://github.com/obsproject/obs-studio/issues/5474